c/tskit/trees.c

```diff
@@ -336,7 +336,7 @@ tsk_treeseq_init_nodes(tsk_treeseq_t *self)
  */
 int TSK_WARN_UNUSED
 tsk_treeseq_init(
-    tsk_treeseq_t *self, tsk_table_collection_t *tables, tsk_flags_t options)
+    tsk_treeseq_t *self, const tsk_table_collection_t *tables, tsk_flags_t options)
```

Comment:
This one is slightly tricky. Currently we're taking a copy of the input tables so it's fine if it's const. But, we plan on making it possible to avoid this copy to make things a bit more efficient. Ultimately, we want to have zero-copy semantics over a table collection all the way through to the Python API (see #53), but this will be a fair bit of work to do with absolute safety.

So, the question is, *if* we're passing in a "frozen" table collection that we just keep a reference to, do we need to update it? I don't think we do, except if it was not indexed. But, it would almost certainly be better to raise an error in this case rather than building the indexes.

I think I've convinced myself that const is good here; @benjeffery, what do you think?

Reply:
Const all the way and deal with any issues that arise in my view. Besides, `tsk_treeseq_t` as a read-only view of `tsk_table_collection_t` is a key architectural decision of the library, we should enforce it.